### PR TITLE
feat: add support for submitting previews to tmc

### PR DESCRIPTION
**Pull request by wmalik**
## What this PR does / why we need it:

* Add flag `--cloud-sync-preview` to `terramate run` command
* Add an interop test
* Create a new package `cloud/previews` to handle the preview creation and add unit tests (using a fake HTTP transport)
* The end-to-end test implementation is skipped for now but there is some test coverage via the unit tests + interop test

## Special notes for your reviewer:

* The functionality to send the stack changeset and stack logs to TMC will be implemented in a follow-up PR

## Does this PR introduce a user-facing change?
```
Yes, a new flag was added to the `terramate run` command.
```

### CODE REVIEW

**Review by sixstone-qq — COMMENTED** [review]
Few comments , let me know what you think of them

File: cloud/previews/previews.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	// loop over all affected stacks, if an item is present in the
+	// previewStacksMap, use the preview status and cmd from there
+	for _, affectedStack := range opts.AffectedStacks {
+		tags := []string{}
+		if len(affectedStack.Tags) > 0 {
+			tags = affectedStack.Tags
+		}
+
+		stack := cloud.PreviewStack{
+			PreviewStatus:   PreviewStatusAffected,
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/previews/previews.go:120]: Pending stacks are a subset of the `affectedStacks`, if there is a stack that's part of the preview run it is updated to the pending status on L132.

File: cmd/terramate/cli/run.go [hunk 2]
```diff
@@ -421,3 +434,65 @@ func (c *cli) loadAllStackEnvs(runs []runContext) (map[prj.Path]runutil.EnvVars,
...
+		fatal("env var GITHUB_EVENT_PATH not found, not generating previews", nil)
+	}
+
+	prUpdatedAt := time.Now().UTC().Unix()
+	if eventPRUpdatedAt := github.GetEventPRUpdatedAt(githubEventPath); eventPRUpdatedAt != nil {
+		prUpdatedAt = eventPRUpdatedAt.Unix()
+	}
+
+	technology := "other"
+	technologyLayer := "other"
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cmd/terramate/cli/run.go]: Fixed in https://github.com/terramate-io/terramate/pull/1436/commits/1a66efdf34c521f5cdec82394f629fb9a1fcb657 (will rebase shortly)

File: cloud/types.go [hunk 3]
```diff
@@ -220,15 +220,16 @@ type (
...
+		Platform          string    `json:"platform"`
+		Repository        string    `json:"repository"`
+		CommitSHA         string    `json:"commit_sha"`
+		Number            int       `json:"number"`
+		Title             string    `json:"title"`
+		Description       string    `json:"description"`
+		URL               string    `json:"url"`
+		Labels            []Label   `json:"labels,omitempty"`
+		Reviewers         Reviewers `json:"reviewers,omitempty"`
+		PlatformUpdatedAt time.Time `json:"platform_updated_at"`
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/types.go]: Fixed in https://github.com/terramate-io/terramate/pull/1436/commits/1a66efdf34c521f5cdec82394f629fb9a1fcb657 (will rebase shortly)

File: cloud/previews/previews.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sixstone-qq) [reply cloud/previews/previews.go:120]: but they are being set to any value, should it be better to limit to pending there only. If not , it would be potentially be a Bad request.

[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/previews/previews.go:119]: makes sense, fixed.

**Review by sixstone-qq — APPROVED** [review]
From my understanding in communication with TMC. All good now

**Review by snakster — COMMENTED** [review]
Looks great, really well done 👍 

One thing I noticed is that your approach for testing with fake transport is new, wouldn't it be more consistent to add it to the e2etest like the rest of the cloud tests, or is there a specific reason you chose this approach?

**Comment by wmalik** [comment]
> Looks great, really well done 👍
> 
> One thing I noticed is that your approach for testing with fake transport is new, wouldn't it be more consistent to add it to the e2etest like the rest of the cloud tests, or is there a specific reason you chose this approach?

@snakster I chose that approach mainly because it was faster to implement. Another reason is that those tests are only testing the `previews` package and hence can be used for fine grained unit tests.

However, I do plan to add end to end tests for this once the MVP has been validated/tested.

**Review by snakster — APPROVED** [review]

File: cloud/previews/previews.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	PreviewStatusAffected = "affected"
+	// PreviewStatusPending is the status for a stack that is selected in a preview run
+	PreviewStatusPending = "pending"
+)
+
+// RunContext is the context for a run
+type RunContext struct {
+	Stack *config.Stack
+	Cmd   []string
+}
```
Comment (i4ki) [inline cloud/previews/previews.go:27]: I don't get why this is leaking here. This type is a helper type of the cli package, also used by the other sync features.

File: cloud/previews/previews.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	AffectedStacks  map[string]*config.Stack
+	OrgUUID         cloud.UUID
+	UpdatedAt       int64
+	Technology      string
+	TechnologyLayer string
+	Repository      string
+	DefaultBranch   string
+	ReviewRequest   *cloud.DeploymentReviewRequest
+	Metadata        *cloud.DeploymentMetadata
+}
```
Comment (i4ki) [inline cloud/previews/previews.go:41]: this is a mix of required and optional fields so I would not call it `...Opts`.
I think it's better to pass the required information explicitly as function parameters and then this `Opt` type can be used for options.

File: cloud/previews/previews.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+	// previewStacksMap, use the preview status and cmd from there
+	for _, affectedStack := range opts.AffectedStacks {
+		tags := []string{}
+		if len(affectedStack.Tags) > 0 {
+			tags = affectedStack.Tags
+		}
+
+		stack := cloud.PreviewStack{
+			PreviewStatus:   PreviewStatusAffected,
+			Repository:      opts.Repository,
```
Comment (i4ki) [inline cloud/previews/previews.go:120]: this is not validated. Is repo okay to be empty?

File: cloud/previews/previews.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+		Technology:      opts.Technology,
+		TechnologyLayer: opts.TechnologyLayer,
+		ReviewRequest:   opts.ReviewRequest,
+		Metadata:        opts.Metadata,
+		Stacks:          []cloud.PreviewStack{},
+	}
+
+	previewStacksMap := map[string]RunContext{}
+	for _, run := range opts.Runs {
+		previewStacksMap[run.Stack.ID] = run
```
Comment (i4ki) [inline cloud/previews/previews.go:107]: TODO: duplication

File: cloud/previews/previews.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+		StackPreviewsByMetaID: stacks,
+	}, nil
+}
+
+func (o CreatePreviewOpts) validate() error {
+	errs := errors.L()
+
+	if len(o.AffectedStacks) == 0 || len(o.Runs) == 0 {
+		errs.Append(errors.E("no affected stacks or runs provided"))
+	}
```
Comment (i4ki) [inline cloud/previews/previews.go:171]: A deployment is created even when no stacks are selected. What was decided in the case of previews? is this never going to happen? is this going to be handled by the GHA script?

File: cloud/previews/previews_test.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+func makeAffectedStacks(num int, isChanged bool) map[string]*config.Stack {
+	affectedStacks := map[string]*config.Stack{}
+	for i := 0; i < num; i++ {
+		affectedStacks[strconv.Itoa(i)] = &config.Stack{
+			Dir:         project.NewPath("/somepath" + strconv.Itoa(i)),
+			ID:          uuid.NewString(),
+			Name:        "stack" + strconv.Itoa(i),
+			Description: "desc of stack" + strconv.Itoa(i),
+			Tags:        []string{"tag1", "tag2"},
+			IsChanged:   isChanged,
```
Comment (i4ki) [inline cloud/previews/previews_test.go:132]: why `IsChanged` is needed here? in the case of deployments, it never matters the reason why the stack was selected.

File: cloud/previews/previews_test.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+			numRunContexts: 3,
+			want: want{
+				numStacksReturned: 3,
+				httpEndpointsCalled: []string{
+					"POST /v1/previews/aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee",
+				},
+				err: nil,
+			},
+		},
+	}
```
Comment (i4ki) [inline cloud/previews/previews_test.go:68]: only the *three stacks* is testing a real case. 
The first is a valid case (should not error), the second never happens because cli ensures the organization is selected. 

I think this testing approach is brittle and is not testing anything..
better do it later with the `testserver` or `interop` package.

File: cloud/types.go [hunk 11]
```diff
@@ -220,23 +220,24 @@ type (
 
 	// DeploymentReviewRequest is the review_request object.
 	DeploymentReviewRequest struct {
```
Comment (i4ki) [inline cloud/types.go:222]: Now that this is reused better just call it `ReviewRequest`. I was confused why this change here if the feature is not related to deployment.

File: cloud/types_previews.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+		}
+		if stack.Path == "" {
+			errs.Append(errors.E("path is required for stack[%d]", i))
+		}
+		if stack.DefaultBranch == "" {
+			errs.Append(errors.E("default_branch is required for stack[%d]", i))
+		}
+		if stack.MetaID == "" {
+			errs.Append(errors.E("meta_id is required required for stack[%d]", i))
+		}
```
Comment (i4ki) [inline cloud/types_previews.go:86]: these validations are repeated because the `Stack` type is not being reused.
The error messages are different and some validations are not in place (`meta_id` not lowercase).

File: cloud/types_previews.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	return errs.AsError()
+}
+
+// Validate the CreatePreviewPayloadRequest object.
+func (r CreatePreviewPayloadRequest) Validate() error {
+	errs := errors.L()
+	if r.Technology == "" {
+		errs.Append(errors.E("technology is required"))
```
Comment (i4ki) [inline cloud/types_previews.go:99]: please follow the error message convention: https://github.com/terramate-io/terramate/blob/main/cloud/types.go#L411-L423

```
errs.Append(errors.E(`missing "technology" field`))
```

no problem changing the way, but it would be good if we return similar messages for similar errors.

File: cmd/terramate/cli/cli.go [hunk 14]
```diff
@@ -2112,15 +2113,18 @@ func (c *cli) friendlyFmtDir(dir string) (string, bool) {
...
 		return nil, err
 	}
 
 	c.gitFileSafeguards(ensureCleanRepo)
 
-	entries := c.filterStacks(report.Stacks)
+	entries := report.Stacks
+	if filterStacks {
+		entries = c.filterStacks(report.Stacks)
+	}
```
Comment (i4ki) [inline cmd/terramate/cli/cli.go:2127]: I see that you want to compute the selected stacks twice, first filtering (by changed, tags, workingdir, cloud status, etc) and then later without some filters but allowing others (changed, cloud status). This seems broken.

Additionally, this logic requires a lot of computation, invoking of lots of git commands, parsing of Terraform modules, etc, then doing this twice is very bad.

File: cmd/terramate/cli/github/event.go [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+		return nil, err
+	}
+
+	var parsedEvent event
+	if err := json.Unmarshal(bytes, &parsedEvent); err != nil {
+		return nil, err
+	}
+
+	return &parsedEvent, nil
+}
```
Comment (i4ki) [inline cmd/terramate/cli/github/event.go:46]: this function returns an error that's always ignored.

File: cmd/terramate/cli/github/event.go [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+// GetEventPRUpdatedAt returns the updated_at field from the file at `eventPath`.
+// The file is expected to be a JSON file containing the GitHub event for a pull
+// request.
+func GetEventPRUpdatedAt(eventPath string) *time.Time {
+	event, err := getEventFromPath(eventPath)
+	if err != nil {
+		return nil
+	}
```
Comment (i4ki) [inline cmd/terramate/cli/github/event.go:19]: I think we should at least add a warning here.

File: cmd/terramate/cli/run.go [hunk 17]
```diff
@@ -421,3 +434,64 @@ func (c *cli) loadAllStackEnvs(runs []runContext) (map[prj.Path]runutil.EnvVars,
...
+	}
+
+	affectedStacksMap := map[string]*config.Stack{}
+	for _, st := range affectedStacks {
+		affectedStacksMap[st.Stack.ID] = st.Stack
+	}
+
+	githubEventPath, ok := os.LookupEnv("GITHUB_EVENT_PATH")
+	if !ok {
+		fatal("env var GITHUB_EVENT_PATH not found, not generating previews", nil)
```
Comment (i4ki) [inline cmd/terramate/cli/run.go:459]: The `sync` features must never abort without executing the provided `run` or `script` command. It must warn and just disable the feature. Check the `c.disableCloudFeatures(err)`.

File: cmd/terramate/cli/run.go [hunk 18]
```diff
@@ -421,3 +434,64 @@ func (c *cli) loadAllStackEnvs(runs []runContext) (map[prj.Path]runutil.EnvVars,
...
+	}
+
+	prUpdatedAt := time.Now().UTC().Unix()
+	if eventPRUpdatedAt := github.GetEventPRUpdatedAt(githubEventPath); eventPRUpdatedAt != nil {
```
Comment (i4ki) [inline cmd/terramate/cli/run.go:463]: it would be nice if we can have a testing env var for injecting the updated at.
```
$ make build/test
$ TM_TEST_PREVIEW_UPDATED_AT=... ./bin/terramate ...
```

File: cmd/terramate/cli/run.go [hunk 19]
```diff
@@ -421,3 +434,64 @@ func (c *cli) loadAllStackEnvs(runs []runContext) (map[prj.Path]runutil.EnvVars,
...
+			Technology:      technology,
+			TechnologyLayer: technologyLayer,
+			Repository:      c.prj.prettyRepo(),
+			DefaultBranch:   c.prj.gitcfg().DefaultBranch,
+			ReviewRequest:   c.cloud.run.reviewRequest,
+			Metadata:        c.cloud.run.metadata,
+		},
+	)
+	if err != nil {
+		fatal("unable to create cloud preview", err)
```
Comment (i4ki) [inline cmd/terramate/cli/run.go:491]: it should never abort.

File: makefiles/unix.mk [hunk 20]
```diff
@@ -38,6 +38,7 @@ test/helper:
 tempdir=$(shell ./bin/helper tempdir)
 test: test/helper build
 # 	Using `terramate` because it detects and fails if the generated files are outdated.
+	./bin/terramate run --no-recursive -- go test -c -tags interop ./cmd/terramate/e2etests/cloud/interop/...
```
Comment (i4ki) [inline makefiles/unix.mk:41]: this must be reverted. The interop tests must be explicit invoked, as they are very slow and require proper environment set up.

**Review by i4ki — COMMENTED** [review]
Some important details need to be addressed.

(I removed most of the nitpick comments because it's already merged but I think we are diverging a lot from the other cloud implementations and it would be good if those changes comes with a good reasoning)

File: makefiles/unix.mk (hunk 20 above)
Comment (i4ki) [reply makefiles/unix.mk:41]: just noticed the `-c` in the `go test`. Why this is needed?

Comment (wmalik) [reply makefiles/unix.mk:41]: since the interop tests have a build tag, they are not built  by `go test` and syntax errors won't be caught by the CI. 
`go test -c` only builds the test (it does not run them).

File: cmd/terramate/cli/run.go (hunk 17 above)
Comment (wmalik) [reply cmd/terramate/cli/run.go:459]: In this case disabling does not make sense because `--cloud-sync-preview` feature requires a timestamp to create a preview.
However, we could handle that case gracefully by using `time.Now()` and removing the `fatal`, what do you think?

File: cloud/types_previews.go [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+	}
+	// CreatePreviewPayloadRequest is the request payload for the creation of
+	// stack deployments.
+	CreatePreviewPayloadRequest struct {
+		UpdatedAt       int64                    `json:"updated_at"`
+		Technology      string                   `json:"technology"`
+		TechnologyLayer string                   `json:"technology_layer"`
+		ReviewRequest   *DeploymentReviewRequest `json:"review_request,omitempty"`
+		Metadata        *DeploymentMetadata      `json:"metadata,omitempty"`
+		Stacks          PreviewStacks            `json:"stacks"`
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/types_previews.go:33]: There is no usecase for `nil` stacks currently. There is a validation for it in the `Validate()` method in case its ever set as `nil`. The API marks this as a required attribute. Could you please link to the case where it is marked as omitempty? I don't see a problem with marking this as `omitempty` though if it's safer for the API.

File: cloud/previews/previews.go [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+	"time"
+
+	"github.com/terramate-io/terramate/cloud"
+	"github.com/terramate-io/terramate/config"
+	"github.com/terramate-io/terramate/errors"
+)
+
+const (
+	// PreviewStatusAffected is the status for a stack that is affected in a PR
+	PreviewStatusAffected = "affected"
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/previews/previews.go:18]: That makes sense, I already addressed it in the previews changeset PR, see https://github.com/terramate-io/terramate/pull/1443/files#diff-7f1a481103c73ed166b25cfecbd5d9fc8b35ede4e2c89dc73d5a373192cba8e5

File: cloud/previews/previews.go (hunk 4 above)
Comment (wmalik) [reply cloud/previews/previews.go:27]: The preview package is unware of the `cli.runContext` type. The `preview.CreatePreview` function needs information about the all the scripts+commands that are part of a `--cloud-sync-preview` invocation, and this type + `[]RunContext` captures that information. It could have a better name though to reduce the confusion.

File: cloud/previews/previews.go (hunk 6 above)
Comment (wmalik) [reply cloud/previews/previews.go:120]: its validated here: https://github.com/terramate-io/terramate/blob/22329f6c6657eb9afe9462f48236bb1053eea6b8/cloud/types_previews.go#L75

but I will change that to use the existing `Stacks` validation to remove the duplication.

File: cloud/previews/previews.go (hunk 8 above)
Comment (wmalik) [reply cloud/previews/previews.go:171]: Good question, I will think about this one and get back to you.

File: cloud/previews/previews.go (hunk 7 above)
Comment (wmalik) [reply cloud/previews/previews.go:107]: Could you please elaborate?

File: cloud/previews/previews.go (hunk 5 above)
Comment (wmalik) [reply cloud/previews/previews.go:41]: main motivation to add that struct was to have a smaller arg list. I would lean towards removing the struct all together and passing all of that information as function args instead of differentiating between optional/mandatory args because it's not meaningful IMO to mark optional/non-optional args. What do you think?

File: cloud/previews/doc.go [hunk 23]
```diff
@@ -0,0 +1,5 @@
+// Copyright 2024 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+// Package previews contains functionality for previews.
+package previews
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/previews/doc.go:5]: Fixed in https://github.com/terramate-io/terramate/pull/1451

File: cloud/previews/previews.go [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+		previewStacksMap[run.Stack.ID] = run
+	}
+
+	// loop over all affected stacks, if an item is present in the
+	// previewStacksMap, use the preview status and cmd from there
+	for _, affectedStack := range opts.AffectedStacks {
+		tags := []string{}
+		if len(affectedStack.Tags) > 0 {
+			tags = affectedStack.Tags
+		}
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/previews/previews.go:116]: Removed in https://github.com/terramate-io/terramate/pull/1451

File: cloud/types_previews.go [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+type (
+	// PreviewStacks is a list of stack objects for the request payload
+	PreviewStacks []PreviewStack
+
+	// PreviewStack represents the stack object of the request payload
+	// type for the creation of stack deployments.
+	PreviewStack struct {
+		PreviewStatus   string   `json:"preview_status"`
+		Repository      string   `json:"repository"`
+		Path            string   `json:"path"`
```
[the remark this answers is not in the document]
Comment (wmalik) [reply cloud/types_previews.go:17]: Makes sense, fixed in https://github.com/terramate-io/terramate/pull/1451

File: cloud/previews/previews_test.go (hunk 9 above)
Comment (wmalik) [reply cloud/previews/previews_test.go:132]: Fixed in https://github.com/terramate-io/terramate/pull/1451

File: cloud/types.go (hunk 11 above)
Comment (wmalik) [reply cloud/types.go:222]: Fixed in https://github.com/terramate-io/terramate/pull/1451

File: cloud/types_previews.go (hunk 12 above)
Comment (wmalik) [reply cloud/types_previews.go:86]: I embedded the `Stack` tupe into the PreviewStack struct and used the existing Stack.Validate() method to remove the duplication in https://github.com/terramate-io/terramate/pull/1451

File: cloud/types_previews.go (hunk 13 above)
Comment (wmalik) [reply cloud/types_previews.go:99]: Makes sense to stay consistent, fixed in https://github.com/terramate-io/terramate/pull/1451

File: cmd/terramate/cli/run.go (hunk 18 above)
Comment (wmalik) [reply cmd/terramate/cli/run.go:463]: `TM_IGNORE_EVENT_PATH`

File: cloud/previews/previews.go (hunk 8 above)
Comment (wmalik) [reply cloud/previews/previews.go:171]: I think the preview should be created even if there are no selected stacks, so I removed the validation.
